src/Storages/ExternalStream/Kafka/KafkaSink.h

```diff
@@ -56,42 +52,62 @@ class KafkaSink final : public SinkToStorage
-        int32_t partition_count,
-        void * rkt_opaque,
-        void * msg_opaque)
-    {
-        /// update partition count
-        auto * sink = static_cast<KafkaSink *>(rkt_opaque);
-        sink->partition_cnt = partition_count;
-
-        auto parition_id = static_cast<Int32>(reinterpret_cast<std::uintptr_t>(msg_opaque));
-        /// This should not really happen because Kafka does not support reducing partitions.
-        /// However, KIP-694 is currently under discussion, so this might heppen in the future.
-        if (parition_id >= partition_count)
-            parition_id = partition_count - 1;
-        return parition_id;
-    }
+    /// Callback for Kafka message delivery report
+    static void onMessageDelivery(rd_kafka_t * /* producer */, const rd_kafka_message_t * msg, void *  /*opaque*/);
+    void onMessageDelivery(const rd_kafka_message_t * msg);
+
+    void addMessageToBatch(char * pos, size_t len);
+
+    /// the number of acknowledgement has been received so far for the current checkpoint period
+    size_t acked() const { return state.acked; }
+    /// the number of errors has been received so far for the current checkpoint period
+    size_t error_count() const { return state.error_count; }
+    /// the number of outstanding messages for the current checkpoint period
+    size_t outstandings() const { return state.outstandings; }
+    /// the last error code received from delivery report callback
+    rd_kafka_resp_err_t lastSeenError() const { return static_cast<rd_kafka_resp_err_t>(state.last_error_code.load()); }
+    /// check if there are no more outstandings (i.e. delivery reports have been recieved
+    /// for all out-go messages, regardless if a message is successfully delivered or not)
+    bool hasNoOutstandings() const { return state.outstandings == state.acked + state.error_count; }
+    /// allows to reset the state after each checkpoint
+    void resetState() { state.reset(); }
 
     static const int POLL_TIMEOUT_MS = 500;
 
-    klog::KafkaPtr producer;
-    klog::KTopicPtr topic;
-    std::unique_ptr<WriteBufferFromKafka> wb;
-    OutputFormatPtr writer;
-    ThreadPool polling_threads;
-    std::atomic_flag is_finished;
     Int32 partition_cnt;
+    bool one_message_per_row;
+    Poco::Logger * log;
+
+    klog::KafkaPtr producer {nullptr, rd_kafka_destroy};
+    klog::KTopicPtr topic {nullptr, rd_kafka_topic_destroy};
+    ThreadPool polling_threads {1};
+    ThreadPool metadata_threads {1};
+    std::atomic_flag is_finished {false};
+
+    std::unique_ptr<WriteBufferFromKafkaSink> wb;
+    OutputFormatPtr writer;
     std::unique_ptr<KafkaStream::ChunkSharder> partitioner;
 
-    Poco::Logger * log;
+    ExpressionActionsPtr message_key_expr;
+    bool delete_message_key_column;
+    size_t message_key_column_pos;
+
+    /// For constructing the message batch
+    std::vector<rd_kafka_message_t> current_batch;
+    std::deque<StringRef> keys_queue;
+    Int32 next_partition;
```

Comment:
we may want inline init the primitive data members like int, bool, size_t etc here.